File: #0664.strange-printer.cpp
class Solution {
public:
    int strangePrinter(string s) {
        
        int size = s.size();
        dp.resize(size, vector<int>(size));
        return strangePrinter(s, 0, size - 1);
    }

private:
    vector<vector<int>> dp;
    int strangePrinter(string &s,
        int left, int right) {

        if (left > right) {
            return 0;
        }

        if (dp[left][right]) {
            return dp[left][right];
        }

        int result = 1 + strangePrinter(s, left + 1, right);
        for (int i = left + 1; i <= right; i++) {
            if (s[i] == s[left]) {
                result = std::min(result,
                    strangePrinter(s, left + 1, i - 1) +
                    strangePrinter(s, i, right));
            }
        }
        return dp[left][right] = result;
    }
};

File: #0664.strange-printer.js
/**
 * @param {string} s
 * @return {number}
 */
var strangePrinter = function(s) {
    const n = s.length;
    const dp = new Array(n).fill(0)
        .map(() => new Array(n).fill(0));

    for (let i = 0; i < n; i++) {
        dp[i][i] = 1;
    }

    for (let i = 2; i <= n; i++) {
        for (let j = 0; j <= n - i; j++) {
            const k = j + i - 1;
            dp[j][k] = i;
            for (let l = j; l < k; l++) {
                dp[j][k] = Math.min(dp[j][k],
                    dp[j][l] + dp[l + 1][k]);
            }

            if (s[j] === s[k]) {
                dp[j][k]--;
            }
        }
    }
    return dp[0][n - 1];
};
